lib/puppet/functions/stdlib/sort_by.rb

```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+# @summary Sort an Array, Hash or String by mapping values through a given block.
+#
+# @example Sort local devices according to their used space.
+#   $facts['mountpoints'].stdlib::sort_by |$m| { $m.dig(1, 'used_bytes') }
+#
+Puppet::Functions.create_function(:'stdlib::sort_by') do
+  # @param ary The Array to sort.
+  # @param block The block for transforming elements of ary.
+  # @return [Array] Returns an ordered copy of ary.
+  dispatch :sort_by_array do
+    param 'Array', :ary
+    block_param 'Callable[1,1]', :block
+  end
+
+  # @param str The String to sort.
+  # @param block The block for transforming elements of str.
+  # @return [String] Returns an ordered copy of str.
+  dispatch :sort_by_string do
+    param 'String', :str
+    block_param 'Callable[1,1]', :block
+  end
+
+  # @param hsh The Hash to sort.
+  # @param block The block for transforming elements of hsh.
+  # @return [Hash] Returns an ordered copy of hsh.
+  dispatch :sort_by_hash1 do
+    param 'Hash', :hsh
+    block_param 'Callable[1,1]', :block
+  end
+
+  # @param hsh The Hash to sort.
+  # @param block The block for transforming elements of hsh.
+  # @return [Hash] Returns an ordered copy of hsh.
+  dispatch :sort_by_hash2 do
+    param 'Hash', :hsh
+    block_param 'Callable[2,2]', :block
+  end
+
+  def sort_by_iterable(iterable, &block)
+    Puppet::Pops::Types::Iterable.asserted_iterable(self, iterable).sort_by(&block)
+  end
+
+  def sort_by_array(ary, &block)
+    sort_by_iterable(ary, &block)
+  end
+
+  def sort_by_string(str, &block)
+    sort_by_iterable(str, &block).join
+  end
+
+  def sort_by_hash1(hsh, &block)
+    sort_by_iterable(hsh, &block).to_h
```

Comment:
This `to_h` makes `sort_by` behave differently that Ruby's `sort_by` which return an array when sorting a Hash (you have to manually `Hash(...)` the result if you want a hash).

I wonder if we should match Ruby or not.  As a Puppet and Ruby developer, I trend to expect some consistency, but obviously the Puppet language is not Ruby so who am I to expect this :shrug:?  Maybe some people have an idea about this?

Reply:
Keeping consistency means that we can simplify the code enormously: only `sort_by_iterable()` is needed, if Ruby can handle blocks with two or a single parameter transparently for us. I went out of my way to distinct between Strings, Arrays and Hashes, precisely because _my_ expectation was that `sort_by()` should return the same data structure that it was fed. If we cannot find an objective argument for either interpretation, we will be unable to decide the "correct" resolution, I guess.

Reply:
> _my_ expectation was that `sort_by()` should return the same data structure that it was fed.

Good point. I think that help: Puppet can currently sort a String and return a String, something Ruby cannot do because `String#sort` does not exist.  So I am fine with returning a Hash when given a Hash because it makes more sense IMHO!